crates/db_connection_pool/src/clickhousepool.rs

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2024 The Spice.ai OSS Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+use std::{collections::HashMap, str::FromStr, sync::Arc};
+
+use async_trait::async_trait;
+use clickhouse_rs::{ClientHandle, Options, Pool};
+use secrets::Secret;
+use snafu::{ResultExt, Snafu};
+
+use crate::dbconnection::{clickhouseconn::ClickhouseConnection, AsyncDbConnection, DbConnection};
+
+use super::{DbConnectionPool, Result};
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("ConnectionPoolError: {source}"))]
+    ConnectionPoolError {
+        source: clickhouse_rs::errors::ConnectionError,
+    },
+
+    #[snafu(display("InvalidConnectionStringError: {source}"))]
+    InvalidConnectionStringError {
+        source: clickhouse_rs::errors::Error,
+    },
+
+    #[snafu(display("ConnectionPoolRunError: {source}"))]
+    ConnectionPoolRunError {
+        source: clickhouse_rs::errors::Error,
+    },
+
+    #[snafu(display("Invalid parameter: {parameter_name}"))]
+    InvalidParameterError { parameter_name: String },
+}
+
+pub struct ClickhouseConnectionPool {
+    pool: Arc<Pool>,
+}
+
+impl ClickhouseConnectionPool {
+    // Creates a new instance of `ClickhouseConnectionPool`.
+    ///
+    /// # Errors
+    ///
+    /// Returns an error if there is a problem creating the connection pool.
+    #[allow(clippy::unused_async)]
+    pub async fn new(
+        params: Arc<Option<HashMap<String, String>>>,
+        secret: Option<Secret>,
+    ) -> Result<Self> {
+        let mut options = Options::default();
+        let mut host: &str = "localhost";
+        let mut user: &str = "default";
+        let mut db: &str = "default";
+        let mut pass: String = "password".to_string();
+        let mut tcp_port: &str = "9000";
```

Comment:
can't these variable just use unwrap_or rather than using mut?

Reply:
I will tidy this up in the separate pr